proposals/flexible-vectors/README.md

```diff
@@ -1,7 +1,1184 @@
+|:------------|:---------------------------------------------------|
+| `vec.v8`    | `vec.i8`                                           |
+| `vec.v16`   | `vec.i16`                                          |
+| `vec.v32`   | `vec.i32`, `vec.f32`                               |
+| `vec.v64`   | `vec.i64`, `vec.f64`                               |
+| `vec.v128`  | `vec.v8x16`, `vec.v16x8`, `vec.v32x4`, `vec.v64x2` |
+
+Mask types are not vector types, and their actual representation differ depending on the target architecture.
+In particular, `vec.m8` and `vec.m16` might have different architectural sizes (or actually be the same).
+In a similar fashion, `vec.v8` and `vec.m8` might also have different architectural sizes (or actually be the same).
+
+## Immediate LaneIdx operands
+
+As the vector length is not known at compile time, it does not make much sense to specify lane indices as immediates.
+
+Therefore, the type `i32` will be used to specify a lane index.
+The range of a lane index is no further constrained, and is intepreted modulo the actual vector length.
+
+If we impose an upper bound on the SIMD width, we can further constraint the range of lane indices.
+For instance, a maximum width of 2048 would allow to store a lane index in a 8-bit integer.
+Similarily, a maximum width of 524288 would allow to store a lane index in a 16-bit integer.
+The would not change the instruction encoding as they all take runtime lane indices which are always `i32`.
+
+## Special Operations
+
+### Vector length
+
+Returns the number of elements.
+Returned value will always be the same during the whole execution of the program.
+
+- `vec.v8.length -> i32`
+- `vec.v16.length -> i32`
+- `vec.v32.length -> i32`
+- `vec.v64.length -> i32`
+- `vec.v128.length -> i32`
+
+### Mask count
+
+Returns the number of active lanes of a `mask`
+
+- `vec.m8.count(m: vec.m8) -> i32`
+- `vec.m16.count(m: vec.m16) -> i32`
+- `vec.m32.count(m: vec.m32) -> i32`
+- `vec.m64.count(m: vec.m64) -> i32`
+- `vec.m128.count(m: vec.m128) -> i32`
+
+```python
+def mask.S.count(m):
+    result = 0
+    for i in range(mask.S.length):
+      if m[i]:
+          result += 1
+    return result
+```
+
+### Mask all active
+
+Returns a mask where all lanes are active
+
+- `vec.m8.all -> vec.m8`
+- `vec.m16.all -> vec.m16`
+- `vec.m32.all -> vec.m32`
+- `vec.m64.all -> vec.m64`
+- `vec.m128.all -> vec.m128`
+
+```python
+def mask.S.all():
+    result = mask.S.New()
+    for i in range(mask.S.length):
+        result[i] = 1
+    return result
+```
+
+### Mask None active
+
+Returns a mask where no lane is active
+
+- `vec.m8.none -> vec.m8`
+- `vec.m16.none -> vec.m16`
+- `vec.m32.none -> vec.m32`
+- `vec.m64.none -> vec.m64`
+- `vec.m128.none -> vec.m128`
+
+```python
+def mask.S.none():
+    result = mask.S.New()
+    for i in range(mask.S.length):
+        result[i] = 0
+    return result
+```
+
+### Mask index CMP
+
+Returns a mask whose active lanes satisfy `(x + laneIdx) CMP n`
```

Comment:
> I would say x is signed, and the addition uses infinite precision, and the comparison is signed and also infinite precision.

To me `x + laneIdx` should not wrap-around as you would expect with LE (or LT) that the mask starts with active lanes and end with inactive ones, no matter the input.
If you want to have first inactive lanes, you would use the GE or GT variants.

To implement the GE/GT on SVE, I think it would be possible to use LT/LE and negation of the result mask.

Reply:
My point is not to discuss the best way to get inactive lanes first, but to emphasize the difference in behaviour between a variant of the operation you propose with wrap-around and the SVE `WHILE*` instructions.

Reply:
I don't quite understand, as I propose the infinite precision version, not the wrap-around (now clarified). AFAIK, this is the same policy as in SVE, and is the one that make the most sense to me as a user.
So I get you want to point-out the difference with the SVE spec, but I don't see where this difference actually is.

Reply:
Yeah, I should have used the past tense (as in `My point was...`). Now that you have clarified the behaviour, I have nothing else to add.